satellite/satellitedb/overlaycache.go

```diff
@@ -902,6 +902,82 @@ func (cache *overlaycache) UpdateUptime(ctx context.Context, nodeID storj.NodeID
 	return getNodeStats(dbNode), Error.Wrap(tx.Commit())
 }
 
+// AllPieceCounts returns a map of node IDs to piece counts from the db.
+// NB: a valid, partial piece map can be returned even if node ID parsing error(s) are returned.
+func (cache *overlaycache) AllPieceCounts(ctx context.Context) (_ map[storj.NodeID]int, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	// NB: `All_Node_Id_Node_PieceCount_By_PieceCount_Not_Number` selects node
+	// ID and piece count from the nodes table where piece count is not zero.
+	rows, err := cache.db.All_Node_Id_Node_PieceCount_By_PieceCount_Not_Number(ctx)
+	if err != nil {
+		return nil, Error.Wrap(err)
+	}
+
+	pieceCounts := make(map[storj.NodeID]int)
+
+	nodeIDErrs := errs.Group{}
+	for _, row := range rows {
+		nodeID, err := storj.NodeIDFromBytes(row.Id)
+		if err != nil {
+			nodeIDErrs.Add(err)
+			continue
+		}
+		pieceCounts[nodeID] = int(row.PieceCount)
+	}
+	return pieceCounts, nodeIDErrs.Err()
+}
+
+func (cache *overlaycache) UpdatePieceCounts(ctx context.Context, pieceCounts map[storj.NodeID]int) (err error) {
+	defer mon.Task()(&ctx, &pieceCounts)(&err)
```

Comment:
Ideally, we should pass `pieceCounts`, however, I think that's too much to register for the call considering that the maybe quite big.

My suggestion was to pass `len(pieceCounts)`.

By the way, the first parameter `ctx` must be a pointer, but the following ones should be what they are passed to the function so in this case, it should be directly `pieceCounts`, there is no need to pass a pointer.

Reply:
Thanks for the explanation. I considered the large size and was planning on having a conversation with @benjaminsirb today about what we want to persist.

Reply:
@benjaminsirb will tell you the requirements from a data science point of view.

In my case, I assess the tracing data for having more introspection about how the system behaves.

For this specific case, having the length of the map give me a clue of the time that the function can take to return.

For example, if the function takes sometimes quite a few time and sometimes returns quickly, knowing the length gives a clue of the reason why; if it takes so long and the length is big, then it makes sense, but it the function takes so long but the length is small, then we would know that we may have a bottleneck in the database.

Reply:
Yea, that makes sense. I think that metric might make more sense to track in a separate monkit call as it seems the convention is to optionally pass the function's arguments to `mon.Task(...)`.

My thought about why it might be useful to track the entire piece counts is to see how they change over time. We can get an instantaneous view of what the piece counts look like at any time because it's in the DB, but I thought it might be interesting to see how this changes over time.  I wasn't sure though if 1) this is that actually useful or 2) are we already doing this somewhere else and/or should we be.

Reply:
I just removed the second argument from the `mon.Task` call for now and added a todo for now.

Reply:
I just removed the second argument from the `mon.Task` call and added a todo for now.

Reply:
I cannot say about monkit because I haven't still used to analyze the data that it collects.

In other systems/libraries that I used for tracing, I traced function parameters and return values because I wanted to have as much information of each call to the functions in order to compare their execution times and in case of an unexpected result (e.g. error) know the input which caused it.

In our code, I  have barely seen that we pass parameters to `monk.Task(..)` and I don't know if that's intentionally or unintentionally because the people were not aware of it.

> I think that metric might make more sense to track in a separate monkit call as it seems the convention is to optionally pass the function's arguments to mon.Task(...).

I'm not sure about that. As far as I understand that call is for tracking the time that the function takes and some functions execution time should vary depending on the input.

Regarding the optional parameters, I don't think that the variadic parameters of `mon.Task` is because they are optional. They need to be an empty interface variadic parameter because it's the way that it can generally work for any function call.

Reply:
I'm not a big fan of this `TODOs` but at the same time, I don't like to stop things of moving forward for a long time.

Because I don't know the reason why we don't pass parameters to `mon.Task` and this is actually a nit on the purpose of this branch, I'm good to approve the PR.

Reply:
I could absolutely be mistaken about this convention, and I agree that we don't seem to be doing it regularly afaik; however, this is what lead me to think that it's a thing: https://godoc.org/gopkg.in/spacemonkeygo/monkit.v2#Scope.Task.